### PR TITLE
[mqtt][mqtt.homeassistant] #5632 Homeassistant improvements

### PR DESCRIPTION
**Pull request by yozik04**
Working on zigbee2mqtt support.

New thingID needs to include nodeId.

Replace invalid char sequences in unique_id with underscore.

### CODE REVIEW

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/AbstractComponent.java [hunk 1]
```diff
@@ -77,6 +77,7 @@ public AbstractComponent(CFactory.ComponentConfiguration componentConfiguration,
         if (groupId == null || StringUtils.isBlank(groupId)) {
             groupId = this.haID.getFallbackGroupId();
         }
+        groupId = groupId.replaceAll("[^A-Za-z0-9_-]+", "_");
```
Comment (davidgraeff) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/AbstractComponent.java]: Not sure about this. Might cause ambiguous channel IDs. I would rather use `URLEncoder.encode` as soon as core has merged my PR and allows the percent character.

Comment (yozik04) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/AbstractComponent.java]: ok

Comment (yozik04) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/AbstractComponent.java]: But what about "." ?

Comment (davidgraeff) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/AbstractComponent.java]: Hm. url encode will probably leave dots untouched that's true. Would hurt to allow it for Core IDs as well, I guess. If that's not possible it should be escaped by %DOT or so. Just replacing characters is no solution, someone could have mqtt channels like "my.channel" and "my_channel".

Comment (yozik04) [reply bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/AbstractComponent.java]: According to `URLEncoder.encode` doc, it does not touch [a-zA-Z0-9._*-]. All others are replaced. So let's url encode and replace "." with "%DOT"?

**Comment by yozik04** [comment]
Pushed as discussed.

**Comment by yozik04** [comment]
How can I update the binding to the latest snapshot?
in karaf:
```
openhab> bundle:update org.openhab.binding.mqtt.homeassistant https://openhab.jfrog.io/openhab/online-repo-snapshot/2.5/org/openhab/addons/bundles/org.openhab.binding.mqtt.homeassistant/2.5.0-SNAPSHOT/org.openhab.binding.mqtt.homeassistant-2.5.0-SNAPSHOT.jar
```

Does not do anything. Still:
```
openhab> bundle:list -s | grep mqtt
286 x Active x  80 x 1.2.1                 x org.eclipse.paho.client.mqttv3
287 x Active x  80 x 2.5.0.201905290429    x org.openhab.binding.mqtt
288 x Active x  81 x 2.5.0.201905290429    x org.openhab.binding.mqtt.generic
289 x Active x  82 x 2.5.0.201905290429    x org.openhab.binding.mqtt.homeassistant
290 x Active x  82 x 2.5.0.201905290429    x org.openhab.binding.mqtt.homie
291 x Active x  80 x 2.5.0.201905191703    x org.openhab.core.io.transport.mqtt
```

Deleting
```
userdata/tmp/mvn/org/openhab/addons/bundles/org.openhab.binding.mqtt.homeassistant/2.5.0-SNAPSHOT/*
```
Does not help.

**Comment by davidgraeff** [comment]
I never used "bundle:update". Usually you stop and uninstall, and then install again. And for installing I usually drop the freshly build jar into the addons dir.

**Comment by yozik04** [comment]
Ahh, ignore. there is still old version in the repo....

**Comment by davidgraeff** [comment]
Core unfortunately has not accepted the percent sign, for backwards compatibility reasons with script files. Instead they want me to use encode/decode methods of the AbstractUID class (not yet merged).

I'll change that line as soon as core has merged the PR.

**Pull request merged**